storage_service/locations/models/s3.py

```diff
@@ -21,68 +22,114 @@
+    s3_bucket = models.CharField(max_length=64,
+        verbose_name=_('S3 Bucket'),
+        blank=True,
+        help_text=_('S3 Bucket Name'))
 
-    def __init__(self, *args, **kwargs):
-        super(S3, self).__init__(*args, **kwargs)
-        self._client = None
-        self._resource = None
+    def _ensure_bucket_exists(self):
+        client = self.s3_resource.meta.client
+        try:
+            client.head_bucket(Bucket=self.bucket_name)
+        except botocore.exceptions.ClientError as e:
+            # If a client error is thrown, then check that it was a 404 error.
+            # If it was a 404 error, then the bucket does not exist.
+            error_code = e.response['Error']['Code']
+            if error_code == '404':
+                client.create_bucket(Bucket=self.bucket_name)
 
     @property
-    def client(self):
-        if self._client is None:
-            self._client = boto3.client(
-                service_name="s3",
-                endpoint_url=self.endpoint_url,
-                aws_access_key_id=self.access_key_id,
-                aws_secret_access_key=self.secret_access_key,
-                region_name=self.region,
-            )
-        return self._client
+    def bucket_name(self):
+        return self.s3_bucket or self.space_id
 
     @property
-    def resource(self):
-        if self._resource is None:
-            self._resource = boto3.resource(
-                service_name="s3",
-                endpoint_url=self.endpoint_url,
-                aws_access_key_id=self.access_key_id,
-                aws_secret_access_key=self.secret_access_key,
-                region_name=self.region,
-            )
+    def s3_resource(self):
+        if not hasattr(self, '_resource'):
+            if self.aws_access_key_id and self.aws_secret_access_key:
+
+                sts_client = boto3.client(
+                    service_name='sts',
+                    aws_access_key_id=self.aws_access_key_id,
+                    aws_secret_access_key=self.aws_secret_access_key,
+                )
+
+                assumed_role = sts_client.assume_role(
+                    RoleArn=self.aws_assumed_role,
+                    RoleSessionName='storage-session',
+                )
+                credentials = assumed_role['Credentials']
+
+                self._resource = boto3.resource(
+                    service_name='s3',
+                    endpoint_url=self.s3_endpoint_url,
+                    region_name=self.s3_region,
+                    aws_access_key_id=credentials['AccessKeyId'],
+                    aws_secret_access_key=credentials['SecretAccessKey'],
+                    aws_session_token=credentials['SessionToken'],
+                )
+            else:
+                self._resource = boto3.resource(
+                    service_name='s3',
+                    endpoint_url=self.s3_endpoint_url,
+                    region_name=self.s3_region,
+                )
```

Comment:
We should revisit this once we have Archivematica running in our infra; I suspect we can simplify a bit. There we don’t pass an access key/secret key; it provides credentials by some mechanism (possibly a special URL that boto3 knows how to request?) which are already configured to use the desired role.

Reply:
as above - access and secret key already present in the s3 plugin, although I've made them optional here so we can support our setup without the access/secret key. I also added the role field as it's needed in our setup when using the access/secret key, although I'm only using it in dev and probably don't need to.